### PR TITLE
Vectorize more nodes (partial fix for #855)

### PR DESCRIPTION
**Pull request by mattdw**
Partial fix for https://github.com/JacquesLucke/animation_nodes/issues/855 – only Number Wiggle, not Euler for now.

I very directly copied the vector wiggle code, so I'm not sure if it's 100% correct or efficient – I'd be keen to hear if there are any problems with it.

There seems like a lot of double calculation around seeds + evolution; I assume that's just to make sure everything has properly unique seeds?

If I get some time I'll come back and do the same thing for the Euler Wiggle node.

### CODE REVIEW

**Comment by mattdw** [comment]
Now with Euler Wiggle. Just Quaternion to do, I guess… looks like random Quaternions are just generated as 4 random 0..1 floats and then normalized?

**Comment by mattdw** [comment]
Quaternion Wiggle support is in, but I'm really confident about it. I notice the non-list version always has w=1.0, which I couldn't replicate when normalizing was happening. Might need a pointer on the maths here (or if my approach is wrong – I also couldn't figure out how to call normalize() from cython, so I've hand-coded it.)

**Comment by mattdw** [comment]
Just bumping this – I think it's a solid improvement, the only question is whether there's a nicer form for random quaternions (they appear to be rotating correctly still.)

**Comment by OmarSquircleArt** [comment]
@mattdw I agree. Sorry for being late on this. We will try reviewing this as soon as possible.
Thanks for contributing!

File: animation_nodes/algorithms/perlin_noise.pyx [hunk 1]
```diff
@@ -26,6 +27,54 @@ def wiggleVectorList(amount, double evolution, amplitude, int octaves, double pe
         values[i] = perlinNoise1D(evolution + i * 354623, persistance, octaves) * _amplitude[i % 3]
     return result
 
+def wiggleFloatList(amount, double evolution, amplitude, int octaves, double persistance):
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/perlin_noise.pyx]: Float Lists in Animation Nodes are actually of type `DoubleList` not `FloatList`, so this function should be edited to reflect that.

File: animation_nodes/algorithms/perlin_noise.pyx [hunk 2]
```diff
@@ -26,6 +27,54 @@ def wiggleVectorList(amount, double evolution, amplitude, int octaves, double pe
...
+    cdef float *values = <float*>result.data
+    cdef Py_ssize_t i
+    for i in range(amount):
+        values[i] = perlinNoise1D(evolution + i * 354623, persistance, octaves) * amplitude
+    return result
+
+def wiggleEulerList(amount, double evolution, amplitude, int octaves, double persistance):
+    cdef EulerList result = EulerList(length = amount)
+    cdef Euler3 *values = <Euler3*>result.data
+    cdef float _amplitude[3]
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/perlin_noise.pyx]: Why is this needed? Just use the components directly.

File: animation_nodes/algorithms/perlin_noise.pyx [hunk 3]
```diff
@@ -26,6 +27,54 @@ def wiggleVectorList(amount, double evolution, amplitude, int octaves, double pe
...
+        values[i].x = perlinNoise1D(evolution + i * 354623, persistance, octaves) * _amplitude[0]
+        values[i].y = perlinNoise1D(evolution + i + 0.33 * 354623, persistance, octaves) * _amplitude[1]
+        values[i].z = perlinNoise1D(evolution + i + 0.66 * 354623, persistance, octaves) * _amplitude[2]
+        values[i].order = 0
+    return result
+
+def wiggleQuaternionList(amount, double evolution, amplitude, int octaves, double persistance):
+    cdef QuaternionList result = QuaternionList(length = amount)
+    cdef Quaternion *values = <Quaternion*>result.data
+    cdef float _amplitude[4]
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/perlin_noise.pyx]: Same as above.

File: animation_nodes/algorithms/perlin_noise.pyx [hunk 4]
```diff
@@ -26,6 +27,54 @@ def wiggleVectorList(amount, double evolution, amplitude, int octaves, double pe
...
+    cdef float _length
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/perlin_noise.pyx]: No need for `_`, just call it `length`.

**Comment by OmarEmaraDev** [comment]
Also, please add your changes to the CHANGELOG file.

File: animation_nodes/algorithms/perlin_noise.pyx (hunk 2 above)
Comment (mattdw) [reply animation_nodes/algorithms/perlin_noise.pyx]: I kept my code as close to wiggleVectorList as possible; it does this, and I assumed it was a cython detail. It does appear to work fine without, so I'll fix up my changes.

Comment (OmarEmaraDev) [reply animation_nodes/algorithms/perlin_noise.pyx]: I think the Vector Wiggle code needed this because of the `_amplitude[i % 3]` part, so it is not needed here.

**Comment by mattdw** [comment]
@OmarEmaraDev thanks for those suggestions – I've implemented them all and tidied up the Quaternion Wiggle logic, which is now behaving more as expected. A note has been added to the changelog.

File: animation_nodes/algorithms/perlin_noise.pyx [hunk 5]
```diff
@@ -26,6 +27,50 @@ def wiggleVectorList(amount, double evolution, amplitude, int octaves, double pe
         values[i] = perlinNoise1D(evolution + i * 354623, persistance, octaves) * _amplitude[i % 3]
     return result
 
+def wiggleDoubleList(amount, double evolution, amplitude, int octaves, double persistance):
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/perlin_noise.pyx]: `amplitude` could be typed as `double` here, I think.

File: animation_nodes/algorithms/perlin_noise.pyx [hunk 6]
```diff
@@ -26,6 +27,50 @@ def wiggleVectorList(amount, double evolution, amplitude, int octaves, double pe
...
+def wiggleEulerList(amount, double evolution, amplitude, int octaves, double persistance):
+    cdef EulerList result = EulerList(length = amount)
+    cdef Euler3 *values = <Euler3*>result.data
+    cdef Py_ssize_t i
+    for i in range(amount):
+        values[i].x = perlinNoise1D(evolution + i * 354623, persistance, octaves) * amplitude[0]
+        values[i].y = perlinNoise1D(evolution + i + 0.33 * 354623, persistance, octaves) * amplitude[1]
+        values[i].z = perlinNoise1D(evolution + i + 0.66 * 354623, persistance, octaves) * amplitude[2]
+        values[i].order = 0
+    return result
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/perlin_noise.pyx]: Sorry for the misunderstanding, but what I meant is that the array of three floats you had before wasn't needed in this case. Instead, just write them as variables outside of the loop:
```python
xAmplitude, yAmplitude, zAmplitude = amplitude[0], amplitude[1], amplitude[2]
```

The same goes for the Quaternion function.
